### PR TITLE
add fixmatch v1

### PR DESCRIPTION
**Pull request by RainFrost1**

### CODE REVIEW

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+  * [5.2.2 基于 Python 预测引擎推理](#522-%E5%9F%BA%E4%BA%8E-python-%E9%A2%84%E6%B5%8B%E5%BC%95%E6%93%8E%E6%8E%A8%E7%90%86)
+  * [5.2.3 基于 C++ 预测引擎推理](#523-%E5%9F%BA%E4%BA%8E-c-%E9%A2%84%E6%B5%8B%E5%BC%95%E6%93%8E%E6%8E%A8%E7%90%86)
+* [5.4 服务化部署](#54-%E6%9C%8D%E5%8A%A1%E5%8C%96%E9%83%A8%E7%BD%B2)
+* [5.5 端侧部署](#55-%E7%AB%AF%E4%BE%A7%E9%83%A8%E7%BD%B2)
+* [5.6 Paddle2ONNX 模型转换与预测](#56-paddle2onnx-%E6%A8%A1%E5%9E%8B%E8%BD%AC%E6%8D%A2%E4%B8%8E%E9%A2%84%E6%B5%8B)
+* [6. 参考资料](#6-%E5%8F%82%E8%80%83%E8%B5%84%E6%96%99)
+
+## 1. 原理介绍
+
+**作者首先通过实验观察到 metric learning 方法学习到的特征在模型稳定后变化总是在一个可接受的范围内，因此提出了利用历史模型产生的特征来近似当前模型的特征，将这些历史特征按FIFO原则放入一个队列中，在一定程度上解耦了候选正负样本的数量与 mini-batch 大小的关系，且由于历史特征不记录梯度，开销相对较小，最终让informative样本的挖掘更加高效，解决了以往的 metric learning 方法只能利用 mini-batch 内部的样本的局限。**
```
Comment (HydrogenSulfate) [inline docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 不相关的介绍未删去、多处markdown格式有问题（如下）
image

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+## 3. 数据准备
+
+**    **在训练及测试的过程中，cifar10数据集会自动下载，请保持联网。如网络问题，则提前下载好[相关数据](https://dataset.bj.bcebos.com/cifar/cifar-10-python.tar.gz)，并在以下命令中，添加如下参数
+
+```
+${cmd} -o DataLoader.Train.dataset.data_file=${data_file} -o DataLoader.UnLabelTrain.dataset.data_file=${data_file} -o DataLoader.Eval.dataset.data_file=${data_file}
+```
+
+**其中：**`${cmd}`为以下的命令，%{data_file}`是下载数据的路径
```
Comment (HydrogenSulfate) [inline docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 可以以FixMatch/FixMatch_cifar10_40.yaml为例，写一个样例命令，然后说明一下命令的哪些部分是可以替换成其它字段的

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+### 5.1 模型评估
+
+**准备用于评估的 **`*.pdparams`模型参数文件，可以使用训练好的模型，也可以使用[4. 模型训练](#4-%E6%A8%A1%E5%9E%8B%E8%AE%AD%E7%BB%83)中保存的模型。
+
+* **以训练过程中保存的 **`latest.pdparams`为例，执行如下命令即可进行评估。
+
+  ```
+  python3.7 tools/eval.py \
+  -c ppcls/configs/ssl/FixMatch/FixMatch_cifar10_40.yaml \
+  -o Global.pretrained_model="./output/WideResNet/best_model_ema.ema"
```
Comment (HydrogenSulfate) [inline docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 去掉`.ema`后缀名

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  # 下载模型
+  cd PaddleClas
+  mkdir pretrained_models
+  cd pretrained_models
+  wget https://paddle-imagenet-models-name.bj.bcebos.com/dygraph/semi_superwised_learning/FixMatch_WideResNet_cifar10_label40.pdparams
+  cd ..
+  # 评估
+  python3.7 tools/eval.py \
+  -c ppcls/configs/ssl/FixMatch/FixMatch_cifar10_40.yaml \
+  -o Global.pretrained_model="pretrained_models/FixMatch_WideResNet_cifar10_label40.pdparams"
```
Comment (HydrogenSulfate) [inline docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 去掉后缀名

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+  ppcls INFO: [Eval][Epoch 0][Iter: 60/157]CELoss: 0.63351, loss: 0.63351, top1: 0.92982, top5: 0.99539, batch_cost: 0.02585s, reader_cost: 0.00015, ips: 2475.86506 images/sec
+  ppcls INFO: [Eval][Epoch 0][Iter: 80/157]CELoss: 0.85084, loss: 0.85084, top1: 0.93191, top5: 0.99576, batch_cost: 0.02578s, reader_cost: 0.00015, ips: 2482.59021 images/sec
+  ppcls INFO: [Eval][Epoch 0][Iter: 100/157]CELoss: 0.04171, loss: 0.04171, top1: 0.93147, top5: 0.99567, batch_cost: 0.02676s, reader_cost: 0.00015, ips: 2391.99053 images/sec
+  ppcls INFO: [Eval][Epoch 0][Iter: 120/157]CELoss: 0.89842, loss: 0.89842, top1: 0.93027, top5: 0.99561, batch_cost: 0.02647s, reader_cost: 0.00015, ips: 2418.24635 images/sec
+  ppcls INFO: [Eval][Epoch 0][Iter: 140/157]CELoss: 0.57866, loss: 0.57866, top1: 0.93107, top5: 0.99568, batch_cost: 0.02678s, reader_cost: 0.00015, ips: 2389.46068 images/sec
+  ppcls INFO: [Eval][Epoch 0][Avg]CELoss: 0.59721, loss: 0.59721, top1: 0.93140, top5: 0.99570
+  ```
+
+  **默认评估日志保存在 **`PaddleClas/output/RecModel/eval.log`中，可以看到我们提供的模型在 cifar10 数据集上的评估指标为top1: 0.93140, top5: 0.99570
```
Comment (HydrogenSulfate) [inline docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 文件夹应该是WideResNet而不是RecModel吧

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+-c ppcls/configs/ssl/FixMatch_cifar10_40.yaml \
+-o -o Global.pretrained_model=output/WideResNet/best_model_ema.ema \
+-o Global.save_inference_dir="./deploy/inference"
+```
+
+#### 5.2.2 基于 Python 预测引擎推理
+
+1. **修改 **`PaddleClas/deploy/configs/inference_cls.yaml`
+
+   * **将 **`infer_imgs:` 后的路径段改为 query 文件夹下的任意一张图片路径（下方配置使用的是 `demo.jpg`图片的路径）
```
Comment (HydrogenSulfate) [inline docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: deploy目录下没有`demo.jpg`？

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+### 5.2 模型推理
+
+#### 5.2.1 推理模型准备
+
+**将训练过程中保存的模型文件转换成 inference 模型，同样以 **`latest.pdparams` 为例，执行以下命令进行转换
+
+```
+python3.7 tools/export_model.py \
+-c ppcls/configs/ssl/FixMatch_cifar10_40.yaml \
+-o -o Global.pretrained_model=output/WideResNet/best_model_ema.ema \
```
Comment (HydrogenSulfate) [inline docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 删去.ema

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+           mean: [0.4914, 0.4822, 0.4465]
+           std: [0.2471, 0.2435, 0.2616]
+           order: hwc
+   PostProcess: null
+   ```
+2. **执行推理命令**
+
+   ```
+   cd ./deploy/
+   python3.7 python/predict_rec.py -c ./configs/inference_rec.yaml
```
Comment (HydrogenSulfate) [inline docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 应该执行的是predict_cls吧

File: ppcls/data/preprocess/__init__.py [hunk 9]
```diff
@@ -44,6 +44,7 @@
 from ppcls.data.preprocess.ops.operators import Padv2
 from ppcls.data.preprocess.ops.operators import RandomRot90
 from .ops.operators import format_data
+from paddle.vision.transforms import Pad as Pad_paddle_vision
```
Comment (HydrogenSulfate) [inline ppcls/data/preprocess/__init__.py:47]: 建议改成`RawPad`

File: ppcls/data/dataloader/cifar.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        elif self._transform_ops_weak and self._transform_ops_strong:
+            image2 = transform(image, self._transform_ops_weak)
+            image2 = image2.transpose((2, 0, 1))
+            image3 = transform(image, self._transform_ops_strong)
+            image3 = image3.transpose((2, 0, 1))
+
+            return (image2, image3, np.int64(label))
+
+
+class Cifar100(Cifar100_paddle):
```
Comment (HydrogenSulfate) [inline ppcls/data/dataloader/cifar.py:71]: Cifar100_paddle改成RawCifar100?

File: ppcls/engine/engine.py [hunk 11]
```diff
@@ -119,6 +124,21 @@ def __init__(self, config, mode="train"):
...
             self.train_dataloader = build_dataloader(
                 self.config["DataLoader"], "Train", self.device, self.use_dali)
+            if self.config["DataLoader"].get('UnLabelTrain', None) is not None:
+                self.unlabel_train_dataloader = build_dataloader(
+                        self.config["DataLoader"], "UnLabelTrain", self.device,
+                        self.use_dali)
+            else:
+                self.unlabel_train_dataloader = None
+
+            # for iter training, such as semi-super wised learning, fixmatch
```
Comment (HydrogenSulfate) [inline ppcls/engine/engine.py]: 这行注释可以删掉，138行已有相关注释

File: ppcls/engine/engine.py [hunk 12]
```diff
@@ -142,8 +162,11 @@ def __init__(self, config, mode="train"):
 
         # build loss
         if self.mode == "train":
-            loss_info = self.config["Loss"]["Train"]
-            self.train_loss_func = build_loss(loss_info)
+            label_loss_info = self.config["Loss"]["Train"]
+            self.train_loss_func = build_loss(label_loss_info)
+            unlabel_loss_info = self.config.get("UnLabelLoss", {}).get("Train",
```
Comment (HydrogenSulfate) [inline ppcls/engine/engine.py:166]: 建议改成if self.config.get("UnLabelLoss", None) is not None，有监督训练下应该没有必要给engine添加unlabel_loss_info这个属性

File: ppcls/engine/engine.py [hunk 13]
```diff
@@ -208,8 +231,9 @@ def __init__(self, config, mode="train"):
         if self.mode == 'train':
             self.optimizer, self.lr_sch = build_optimizer(
                 self.config["Optimizer"], self.config["Global"]["epochs"],
-                len(self.train_dataloader) // self.update_freq,
-                [self.model, self.train_loss_func])
+                len(self.train_dataloader)
+                if self.iter_per_epoch is None else self.iter_per_epoch //
+                self.update_freq, [self.model, self.train_loss_func])
```
Comment (HydrogenSulfate) [inline ppcls/engine/engine.py]: iter_per_epoch在train的时候一定不为None，这里应该不需要判断，直接使用self.iter_per_epoch//self.update_freq应该就可以了

File: ppcls/engine/train/train_fixmatch.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+def interleave(x, size):
+    s = list(x.shape)
+    return x.reshape([-1, size] + s[1:]).transpose(
+        [1, 0, 2, 3, 4]).reshape([-1] + s[1:])
+
+
+def de_interleave(x, size):
+    s = list(x.shape)
+    return x.reshape([size, -1] + s[1:]).transpose(
+        [1, 0, 2]).reshape([-1] + s[1:])
```
Comment (HydrogenSulfate) [inline ppcls/engine/train/train_fixmatch.py:165]: 两个函数没有用到可以删掉？

File: ppcls/optimizer/learning_rate.py [hunk 15]
```diff
@@ -475,3 +474,33 @@ def get_lr(self):
...
+                return float(current_step) / float(
+                    max(1, self.num_warmup_steps))
+            no_progress = float(current_step - self.num_warmup_steps) / \
+                        float(max(1, self.epochs * self.step_each_epoch - self.num_warmup_steps))
+            return max(0., math.cos(math.pi * self.num_cycles * no_progress))
+
+        return lr.LambdaDecay(
+            learning_rate=self.learning_rate,
+            lr_lambda=_lr_lambda,
+            last_epoch=self.last_epoch)
```
Comment (HydrogenSulfate) [inline ppcls/optimizer/learning_rate.py]: 能否基于LRBase类进行合并呢？

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md (hunk 1 above)
Comment (RainFrost1) [reply docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 已经修改

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md (hunk 2 above)
Comment (RainFrost1) [reply docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 已经修改

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md (hunk 3 above)
Comment (RainFrost1) [reply docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 存储模型，就为best_model_ema.ema.pdparams

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md (hunk 4 above)
Comment (RainFrost1) [reply docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 已经修改

File: docs/zh_CN/training/semi_superwised_learning/FixMatch.md (hunk 5 above)
Comment (RainFrost1) [reply docs/zh_CN/training/semi_superwised_learning/FixMatch.md]: 已经修改

File: ppcls/engine/engine.py (hunk 11 above)
Comment (RainFrost1) [reply ppcls/engine/engine.py]: 已经删除

File: ppcls/engine/engine.py (hunk 12 above)
Comment (RainFrost1) [reply ppcls/engine/engine.py:166]: unlabel的loss，若没有的config.UnlabelLoss，就设为None了，这样也可以吧

File: ppcls/engine/engine.py (hunk 13 above)
Comment (RainFrost1) [reply ppcls/engine/engine.py]: 已经修改

File: ppcls/engine/train/train_fixmatch.py (hunk 14 above)
Comment (RainFrost1) [reply ppcls/engine/train/train_fixmatch.py:165]: 这个我建议先保留，虽然没有用到，但是暂时还不清楚，pytorch版本为什么要用这个操作

File: ppcls/optimizer/learning_rate.py [hunk 16]
```diff
@@ -475,3 +474,47 @@ def get_lr(self):
...
+
+class CosineFixmatch(LRBase):
+    """Cosine decay in FixMatch style
+
+    Args:
+        epochs (int): total epoch(s)
+        step_each_epoch (int): number of iterations within an epoch
+        learning_rate (float): learning rate
+        num_warmup_steps (int): the number warmup steps.
+        warmunum_cycles (float, optional): the factor for cosine in FixMatch learning rate. Defaults to 7 / 16.
```
Comment (HydrogenSulfate) [inline ppcls/optimizer/learning_rate.py:487]: warmunum_cycles -> num_cycles?

**Review by HydrogenSulfate — APPROVED** [review]
